sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala

```diff
@@ -163,6 +165,45 @@ class CollationStringExpressionsSuite
     })
   }
 
+  test("UPPER check output type on collated string") {
+    def testUpper(expected: String, collationId: Int, input: String): Unit = {
+      val s = Literal.create(input, StringType(collationId))
+
+      checkEvaluation(Collation(Upper(s)).replacement, expected)
+    }
+
+    testUpper("UTF8_BINARY", 0, "abc")
+    testUpper("UTF8_BINARY_LCASE", 1, "abc")
+    testUpper("UNICODE", 2, "abc")
+    testUpper("UNICODE_CI", 3, "abc")
+  }
+
+  test("LOWER check output type on collated string") {
+    def testLower(expected: String, collationId: Int, input: String): Unit = {
+      val s = Literal.create(input, StringType(collationId))
+
+      checkEvaluation(Collation(Lower(s)).replacement, expected)
+    }
+
+    testLower("UTF8_BINARY", 0, "abc")
+    testLower("UTF8_BINARY_LCASE", 1, "abc")
+    testLower("UNICODE", 2, "abc")
+    testLower("UNICODE_CI", 3, "abc")
+  }
+
+  test("INITCAP check output type on collated string") {
```

Comment:
@uros-db @nikolamand-db @dbatomic @stefankandic @stevomitric @cloud-fan @MaxGekk  I am a bit sceptical about the recent change we introduced in this Suite with UTF8String refactoring. Why are we opting to e2e tests for each expression? Correct me if I am wrong, but these expressions are proven to work e2e with existing tests. Shouldn't we write only unit tests for each expression to test whether these "black boxes" work correctly under different types of collations and reserve e2e tests for multiple stuff, like correctly working from parsing through analysis to execution.

Reply:
The line of reasoning is to cover both unit tests and e2e tests:
1. first use [CollationSupportSuite.java](https://github.com/apache/spark/blob/master/common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java) with a lot of rigorous tests to see whether collation-aware implementation introduced in `CollationSupport.scala` is correct
2. then use [CollationStringExpressionsSuite.scala](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala) with just a couple of SQL tests to verify whether full queries with collation-aware expressions in `stringExpressions.scala` work as expected

I tried to boil this down in: https://issues.apache.org/jira/browse/SPARK-47410

Reply:
It's a good practice to always have end-to-end tests for each function. The test coverage doesn't need to be very high (the unit tests should cover all the corner cases), but we should have one.